contracts/0.8.25/vaults/VaultHubViewerV1.sol

```diff
@@ -0,0 +1,137 @@
+                return hasRole;
+            } catch {
+                return false;
+            }
+        }
+        return false;
+    }
+
+    /// @notice Checks if a given address has a given role on a vault
+    /// @param vault The vault to check
+    /// @param _member The address to check
+    /// @param _role The role to check
+    /// @return True if the address has the role, false otherwise
+    function isHasRole(IVault vault, address _member, bytes32 _role) public view returns (bool) {
+        address owner = vault.owner();
+        if (owner == address(0)) {
+            return false;
+        }
+
+        try IDashboard(owner).hasRole(_role, _member) returns (bool hasRole) {
+                return hasRole;
+        } catch {
+                return false;
+        }
+    }
+
+    /// @notice Returns all vaults owned by a given address
+    /// @param _owner Address of the owner
+    /// @return An array of vaults owned by the given address
+    function vaultsByOwner(address _owner) public view returns (IVault[] memory) {
+        uint256 count = vaultHub.vaultsCount();
+        IVault[] memory vaults = new IVault[](count);
+
+        // Populate the array with the owner's vaults
+        for (uint256 i = 0; i < count; i++) {
+            if (isOwner(vaultHub.vault(i), _owner)) {
+                vaults[i] = vaultHub.vault(i);
+            }
+        }
+
+        return _filterNonZeroVaults(vaults);
+    }
+
+    /// @notice Returns all vaults with a given role on a given address
+    /// @param _role Role to check
+    /// @param _member Address to check
+    /// @return An array of vaults with the given role on the given address
+    function vaultsByRole(bytes32 _role, address _member) public view returns (IVault[] memory) {
+        uint256 count = vaultHub.vaultsCount();
+        IVault[] memory vaults = new IVault[](count);
+
+        for (uint256 i = 0; i < count; i++) {
+            if (isHasRole(vaultHub.vault(i), _member, _role)) {
+                vaults[i] = vaultHub.vault(i);
+            }
+        }
+
+        return _filterNonZeroVaults(vaults);
+    }
+
+    /// @notice Filters out zero address vaults from an array
+    /// @param vaults Array of vaults to filter
+    /// @return An array of non-zero vaults
+    function _filterNonZeroVaults(IVault[] memory vaults) internal pure returns (IVault[] memory) {
+        uint256 nonZeroLength = 0;
+        for (uint256 i = 0; i < vaults.length; i++) {
+            if (address(vaults[i]) != address(0)) {
+                nonZeroLength++;
+            }
+        }
+        IVault[] memory nonZeroVaults = new IVault[](nonZeroLength);
+        uint256 index = 0;
+        for (uint256 i = 0; i < vaults.length; i++) {
+            if (address(vaults[i]) != address(0)) {
+                nonZeroVaults[index] = vaults[i];
+                index++;
+            }
+        }
+        return nonZeroVaults;
+    }
```

Comment:
```suggestion
    function _filterNonZeroVaults(IVault[] memory _vaults, uint256 _validCount) internal pure returns (IVault[] memory filtered)
    {
        filtered = new IVault[](_validCount);
        for (uint256 i = 0; i < _validCount; i++) {
            filtered[i] = _vaults[i];
        }
    }
```

Reply:
It might not be necessary to have a separate function; instead, it could be integrated into the existing ones.

Reply:
fix applied